README.md
# Hacking concurrency programming with an example

This repo holds solutions to the hacking concurrency post in [hackerclub](https://hackerclub.io/hacking-concurrency/).

Go code implements a single threaded solution, a locking solution and a queueing solution.

Clojure code implements a transactional based solution.

## Running

Clojure  
Leiningen is a dependency, please install in [here](https://leiningen.org/#install).

```shell
$ cd clj
$ lein run
```

Go
```shell
$ cd go
$ go run tickets.go

$ cd locking
$ go run tickets.go

$ cd queueing
$ go run tickets.go
```

You can also try it in this [repl.it](https://repl.it/@rodrigorahal/Tickets).

clj/CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-01-10
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-01-10
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/tickets/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/tickets/compare/0.1.0...0.1.1


project.clj
(defproject tickets "0.1.0-SNAPSHOT"
  :description "FIXME: write description"
  :url "http://example.com/FIXME"
  :license {:name "EPL-2.0 OR GPL-2.0-or-later WITH Classpath-exception-2.0"
            :url "https://www.eclipse.org/legal/epl-2.0/"}
  :dependencies [[org.clojure/clojure "1.10.1"]
                 [org.clojure/tools.logging "1.1.0"]
                 [org.clojure/core.async "1.3.610"]
                 [log4j/log4j "1.2.17" :exclusions [javax.mail/mail
                                                    javax.jms/jms
                                                    com.sun.jmdk/jmxtools
                                                    com.sun.jmx/jmxri]]]
  :main ^:skip-aot tickets.transactions
  :target-path "target/%s"
  :profiles {:uberjar {:aot :all
                       :jvm-opts ["-Dclojure.compiler.direct-linking=true"]}})


go/tickets.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"os"
)

const (
	StandardSeatsCapacity int = 45
	PremiumSeatsCapacity  int = 15
)

type Customer struct {
	ID       int  `json:"id"`
	Upgrades bool `json:"upgrades"`
}

type Seats struct {
	Standard map[int]bool
	Premium  map[int]bool
}

func New() *Seats {
	s := make(map[int]bool, StandardSeatsCapacity)
	p := make(map[int]bool, PremiumSeatsCapacity)
	return &Seats{Standard: s, Premium: p}
}

func (s *Seats) Book(c Customer) {
	if len(s.Standard) < StandardSeatsCapacity {
		s.Standard[c.ID] = true
	}
}

func (s *Seats) Upgrade(c Customer) {
	if c.Upgrades && len(s.Premium) < PremiumSeatsCapacity && s.Standard[c.ID] {
		delete(s.Standard, c.ID)
		s.Premium[c.ID] = true
	}
}

func main() {
	var customers []Customer

	file, _ := os.Open("input.json")
	bytes, _ := ioutil.ReadAll(file)
	json.Unmarshal(bytes, &customers)

	seats := New()

	for _, customer := range customers {
		seats.Book(customer)
		seats.Upgrade(customer)
	}

	fmt.Printf("Booked %d Standard seats\n", len(seats.Standard))
	fmt.Printf("Booked %d Premium seats\n", len(seats.Premium))
	fmt.Println(seats)
}


src/tickets/core.clj
(ns tickets.core
  (:require [tickets.input :as input])
  (:gen-class))

(def STANDARD_SEATS_CAPACITY 45)
(def PREMIUM_SEATS_CAPACITY 15)

(def seats (atom {:standard #{}
                  :premium #{}}))

(defn book [seats customer]
  (when (< (count (:standard @seats)) STANDARD_SEATS_CAPACITY)
    (swap! seats update-in [:standard] conj (:id customer))))

(defn offer-upgrade [seats customer]
  (when (and
         (:upgrades customer)
         (< (count (:premium @seats)) PREMIUM_SEATS_CAPACITY))
    (swap! seats update-in [:standard] disj (:id customer))
    (swap! seats update-in [:premium] conj (:id customer))))

(defn sell []
  (let [customers (input/read-json-file "input.json")
        seats (atom {:standard #{}
                     :premium #{}})]

    (doseq [customer customers]
      (book seats customer)
      (offer-upgrade seats customer))

    (println seats)
    (println "Booked " (count (:standard @seats)) "standard seats")
    (println "Booked " (count (:premium @seats)) "standard seats")))

(defn -main
  "I don't do a whole lot ... yet."
  [& args]
  (sell))


clj/src/tickets/transactions.clj
(ns tickets.transactions
  (:require
   [clojure.core.async :refer [go <! >! >!! go-loop chan close! onto-chan]]
   [clojure.tools.logging :as log]
   [tickets.input :as input])
  (:gen-class))

(def STANDARD_SEATS_CAPACITY 45)
(def PREMIUM_SEATS_CAPACITY 15)

(defn book [standard customer]
  (if (< (count @standard) STANDARD_SEATS_CAPACITY)
    (alter standard conj (:id customer))
    standard))

(defn upgrade [standard premium customer]
  (if
   (and (:upgrades customer)
        (< (count @premium) PREMIUM_SEATS_CAPACITY)
        (contains? @standard (:id customer)))
    (do
      (alter standard disj (:id customer))
      (alter premium conj (:id customer)))
    [standard premium]))

(defn start-cashier [id queue standard premium]
  (go-loop []
    (let [customer (<! queue)]
      (when customer
        (dosync
         (log/info (format "Customer %d was attended by Cashier %d" (:id customer) id))
         (book standard customer)
         (upgrade standard premium customer))
        (recur)))))

(defn open-box-office [number-of-cashiers]
  (let [customers (input/read-json-file "input.json")
        standard (ref #{})
        premium (ref #{})
        queue (chan)]

    (dotimes [i number-of-cashiers]
      (start-cashier i queue standard premium))

    (doseq [customer customers]
      (>!! queue customer))

    (println "Booked " (count @standard) "standard seats")
    (println "Booked " (count @premium) "premium seats")
    (println @standard)
    (println @premium)))

(defn -main
  [& args]
  (open-box-office 3))
